File: CS/CustomCachedDocumentSourceSerialization/CategoriesReport.cs
using System;
using System.Collections.Generic;
using DevExpress.XtraReports.UI;

namespace CustomCachedDocumentSourceSerialization {
    public partial class CategoriesReport : DevExpress.XtraReports.UI.XtraReport {
        Dictionary<int, List<int>> pageAdditionalData = new Dictionary<int, List<int>>();

        public CategoriesReport() {
            InitializeComponent();
            this.PrintingSystem.XlSheetCreated += PrintingSystem_XlSheetCreated;
            this.PrintingSystem.AddService(typeof(CustomPageDataService), new CustomPageDataService(pageAdditionalData));
        }

        void PrintingSystem_XlSheetCreated(object sender, DevExpress.XtraPrinting.XlSheetCreatedEventArgs e) {
            var serviceProvider = PrintingSystem as IServiceProvider;
            CustomPageDataService customPageDataService = null;
            if (serviceProvider != null && (customPageDataService = serviceProvider.GetService(typeof(CustomPageDataService)) as CustomPageDataService) != null) {
                customPageDataService.PrintingSystem_XlSheetCreated(sender, e);
            }
        }

        private void CategoryId_PrintOnPage(object sender, PrintOnPageEventArgs e) {
            var cell = sender as XRTableCell;
            int categoryNumber;
            if (cell != null && int.TryParse(cell.Text, out categoryNumber)) {
                if (!pageAdditionalData.ContainsKey(e.PageIndex)) {
                    pageAdditionalData[e.PageIndex] = new List<int>();
                }
                pageAdditionalData[e.PageIndex].Add(categoryNumber);
            }
        }
    }
}

File: VB/CustomCachedDocumentSourceSerialization/CategoriesReport.vb
Imports System
Imports System.Collections.Generic
Imports DevExpress.XtraReports.UI

Namespace CustomCachedDocumentSourceSerialization
	Partial Public Class CategoriesReport
		Inherits DevExpress.XtraReports.UI.XtraReport

		Private pageAdditionalData As New Dictionary(Of Integer, List(Of Integer))()

		Public Sub New()
			InitializeComponent()
			AddHandler PrintingSystem.XlSheetCreated, AddressOf PrintingSystem_XlSheetCreated
			Me.PrintingSystem.AddService(GetType(CustomPageDataService), New CustomPageDataService(pageAdditionalData))
		End Sub

		Private Sub PrintingSystem_XlSheetCreated(ByVal sender As Object, ByVal e As DevExpress.XtraPrinting.XlSheetCreatedEventArgs)
			Dim serviceProvider = TryCast(PrintingSystem, IServiceProvider)
			Dim customPageDataService As CustomPageDataService = Nothing
			customPageDataService = TryCast(serviceProvider.GetService(GetType(CustomPageDataService)), CustomPageDataService)
'INSTANT VB WARNING: An assignment within expression was extracted from the following statement:
'ORIGINAL LINE: if (serviceProvider != null && (customPageDataService = TryCast(serviceProvider.GetService(typeof(CustomPageDataService)), CustomPageDataService)) != null)
			If serviceProvider IsNot Nothing AndAlso customPageDataService IsNot Nothing Then
				customPageDataService.PrintingSystem_XlSheetCreated(sender, e)
			End If
		End Sub

		Private Sub CategoryId_PrintOnPage(ByVal sender As Object, ByVal e As PrintOnPageEventArgs) Handles xrTableCell2.PrintOnPage
			Dim cell = TryCast(sender, XRTableCell)
			Dim categoryNumber As Integer = Nothing
			If cell IsNot Nothing AndAlso Integer.TryParse(cell.Text, categoryNumber) Then
				If Not pageAdditionalData.ContainsKey(e.PageIndex) Then
					pageAdditionalData(e.PageIndex) = New List(Of Integer)()
				End If
				pageAdditionalData(e.PageIndex).Add(categoryNumber)
			End If
		End Sub
	End Class
End Namespace
